### PR TITLE
[AMDGPU][Attributor] Add an option to turn on internalization

### PR DESCRIPTION
**Pull request by shiltian**
This PR adds an opt-in feature to internalize the module before running `AMDGPUAttributor` in LTO pipeline.

The closed-world assumption is not always true for AMDGPU, even at LTO post-link stage, because AMDGPU does support relocatable linking. As a consequence, we can't assume we have a closed-world view at IR level. In this case, if we have an indirect function call inside an externally visible function, whose argument determines the potential callee, we can't simply assume the potential callee would for sure come from the module. Even if we force the closed-world assumption at our own discretion, this doesn't make the situation better because `AAPotentialValues` can't work as expected if a function is externally visible. The closed-world assumption can only deem all functions whose addresses are taken as potential callees.

As shown in the newly added test case, there is an indirect function call inside `@helper`, whose only argument is used as the function pointer. There are three functions whoese address are taken in the module. Without closed-world assumption, we can't optimize the indirect call. With closed-world assumption, all the three functions are taken as potential callee, and we can't deduce it further because `AAPotentialValues` on `%fn` can't work as expected due to the external visibility of `@helper`. With internalization, an internalized version of `@helper` is created, and then everything pans out, thus there are only two potential callees.

Since this is an opt-in feature, it doesn't change any behavior by default, and whoever uses it is at their own discretion.

### CODE REVIEW

**Comment by shiltian** [comment]
* **#108420** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/108420?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by arsenm — COMMENTED** [review]
We already have amdgpu-internalize-symbols, and this looks like it's approximately the same thing. I would like to remove amdgpu-internalize-symbols, not add a second version of it.

I thought the whole point of moving to the LTO pipeline was that the default internalization logic should work.

File: llvm/test/CodeGen/AMDGPU/indirect-call-internalization.ll [hunk 1]
```diff
@@ -0,0 +1,202 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --include-generated-funcs --version 5
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -passes=amdgpu-attributor %s | FileCheck --check-prefixes=EXT %s
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -passes="amdgpu-attributor<closed-world>" %s | FileCheck --check-prefixes=CW %s
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -passes="amdgpu-attributor<internalization>" %s | FileCheck --check-prefixes=INT %s
+
+@G = global i32 0, align 4
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/indirect-call-internalization.ll:6]: don't use globals in addrspace(0)

**Comment by shiltian** [comment]
This is not needed since we already have the internalization pass enabled (via a command line argument passed by the driver) in the pipeline.

**Pull request closed without merging**